### ISSUE TITLE
Regex bugs

### ISSUE THREAD

**Issue by tovine** [riscv-software-src/riscv-config#104]
Having worked a bit too much with regular expressions in the past I've spotted a few issues with the regex, first one:
`[IE]+` at the beginning - this one matches _one or more_ of the characters within the brackets:
![image](https://user-images.githubusercontent.com/8265757/192024434-c8c569d7-6344-4188-824a-6a9c8f1aab63.png)

What you want instead there is just `^RV(32|64|128)[IE]` - that will match _just one of them_, exactly once.

Then the next issue: `[ABCDFGJKLMNPQSTUVX]*` - this group _also_ includes the letters `I` and `E`, so they should be removed as they're already matched by the first one (and should be exclusive)

And third: the current regex requires all the Z-extensions to be in a given order when not the first (but the first one can be whichever).

I've rewritten it to be a bit simpler and work more as (I think) it's intended: 
```regex
"^RV(32|64|128)[IE][ABCDFGJKLMNPQSTUVX]*((Zicsr|Zifencei|Zihintpause|Zam|Ztso|Zkne|Zknd|Zknh|Zkse|Zksh|Zkg|Zkb|Zkr|Zks|Zkn|Zbc|Zbb|Zbp|Zbm|Zbe|Zbf)(_(Zicsr|Zifencei|Zihintpause|Zam|Ztso|Zkne|Zknd|Zknh|Zkse|Zksh|Zkg|Zkb|Zkr|Zks|Zkn|Zbc|Zbb|Zbp|Zbm|Zbe|Zbf))*){,1}$"
```
Working example:
![image](https://user-images.githubusercontent.com/8265757/192027622-c1bc2a7b-0c18-4ce2-a52d-4c2256f1458f.png)

Negative examples (input errors which are correctly not matched):
Extra "B" after starting _Z extensions
![image](https://user-images.githubusercontent.com/8265757/192028146-843404b5-740b-4b74-b7b8-ac2f39d8bc35.png)
Starting with _ before the first Z extension
![image](https://user-images.githubusercontent.com/8265757/192028217-7d5643d5-fefd-4b3a-b6c3-c5b7370dcb0d.png)
Missing underscore between two consecutive Z extensions
![image](https://user-images.githubusercontent.com/8265757/192028482-91f38c3b-dc34-43a7-9e6b-7b7e922546fa.png)

I'll be happy to open a pull request with this new regex if you want (or you can just snip it from this issue post).
I'm already a RVI member (through my employer) if that's a concern 🙂

**Comment by tovine** [comment]
There are most likely several people in this community who are even better at regex than me, they're welcome to optimize my expression even further, but I believe this is a good start at least - ideally the goal would be to only need to specify each extension once, but that's beyond the limit of my regex knowledge.

**Comment by neelgala** [comment]
@tovine I agree with your solution. Please raise a PR. The changes will have to go [here](https://github.com/riscv-software-src/riscv-config/blob/568354139529c60628a417c6a584c6a6a12ee858/riscv_config/constants.py#L10)

**Comment by tovine** [comment]
I'm working on it now, and for the constants.py file I'd honestly like to propose an alternative method of listing all the Z extensions if you'd accept it.
Only downside to this would be that then it's not the same between the python and the YAML version ([schema_isa.yaml](https://github.com/riscv-software-src/riscv-config/blob/master/riscv_config/schemas/schema_isa.yaml)), although I can't find the referenced part in [new-extensions](https://github.com/riscv-software-src/riscv-config/blob/master/docs/source/new-extensions.rst#modifications-in-schema_isayaml) anymore.

Is the YAML file no longer used?

**Comment by tovine** [comment]
Here's my updated suggestion: https://github.com/riscv-software-src/riscv-config/compare/master...tovine:riscv-config:isa_regex_improvements-issue_104#diff-c237b89a7835c67fa049bd6aab6f82dc8be925fb5252430a2a35020231c20059

**Comment by neelgala** [comment]
> I'm working on it now, and for the constants.py file I'd honestly like to propose an alternative method of listing all the Z extensions if you'd accept it. Only downside to this would be that then it's not the same between the python and the YAML version ([schema_isa.yaml](https://github.com/riscv-software-src/riscv-config/blob/master/riscv_config/schemas/schema_isa.yaml)), although I can't find the referenced part in [new-extensions](https://github.com/riscv-software-src/riscv-config/blob/master/docs/source/new-extensions.rst#modifications-in-schema_isayaml) anymore.
> 
> Is the YAML file no longer used?

The ISA node in the YAML file no longer holds the regex. It simply calls the capture_isa_specifics which performs the regex and other checks using the isa_validator.py package. This was done so that other tools trying to parse the isa-string can simply import this file and perform required checks - as opposed to extracting the regex from the yaml and replicating the functions at their end again..

**Comment by neelgala** [comment]
closed in #105
